crates/red_knot/src/salsa_db/semantic/ast_ids.rs

```diff
@@ -0,0 +1,406 @@
+    statements: IndexVec<StatementId, AstNodeRef<Stmt>>,
+    statements_map: FxHashMap<NodeKey, StatementId>,
+
+    deferred: VecDeque<&'a [Stmt]>,
+}
+
+impl AstIdsBuilder<'_> {
+    fn finish(mut self) -> AstIds {
+        self.expressions.shrink_to_fit();
+        self.expressions_map.shrink_to_fit();
+        self.statements.shrink_to_fit();
+        self.statements_map.shrink_to_fit();
+
+        AstIds {
+            expressions: self.expressions,
+            expressions_map: self.expressions_map,
+            statements: self.statements,
+            statements_map: self.statements_map,
+        }
+    }
+}
+
+#[allow(unsafe_code)]
+impl<'a> PreorderVisitor<'a> for AstIdsBuilder<'a> {
+    fn visit_stmt(&mut self, stmt: &'a Stmt) {
+        let statement_id = self
+            .statements
+            .push(unsafe { AstNodeRef::new(self.parsed.clone(), stmt) });
+        self.statements_map
+            .insert(NodeKey::from_node(stmt.into()), statement_id);
+
+        match stmt {
+            Stmt::FunctionDef(StmtFunctionDef {
+                parameters,
+                body,
+                decorator_list,
+                returns,
+                type_params,
+                is_async: _,
+                name: _,
+                range: _,
+            }) => {
+                for decorator in decorator_list {
+                    self.visit_decorator(decorator);
+                }
+
+                if let Some(type_params) = type_params {
+                    self.visit_type_params(type_params);
+                }
+
+                self.visit_parameters(parameters);
+
+                for expr in returns {
+                    self.visit_annotation(expr);
+                }
+
+                self.deferred.push_back(body);
+            }
+
+            Stmt::ClassDef(StmtClassDef {
+                arguments,
+                body,
+                decorator_list,
+                type_params,
+                name: _,
+                range: _,
+            }) => {
+                for decorator in decorator_list {
+                    self.visit_decorator(decorator);
+                }
+
+                if let Some(type_params) = type_params {
+                    self.visit_type_params(type_params);
+                }
+
+                if let Some(arguments) = arguments {
+                    self.visit_arguments(arguments);
+                }
+
+                self.deferred.push_back(body);
+            }
+
+            _ => preorder::walk_stmt(self, stmt),
+        }
+    }
+
+    fn visit_expr(&mut self, expr: &'a Expr) {
+        let expression_id = self
+            .expressions
+            .push(unsafe { AstNodeRef::new(self.parsed.clone(), expr) });
+
+        self.expressions_map
+            .insert(NodeKey::from_node(expr.into()), expression_id);
+
+        preorder::walk_expr(self, expr);
+    }
+}
+
+#[derive(Clone)]
+struct AstNodeRef<T> {
+    _parsed: Arc<Parsed<ModModule>>,
```

Comment:
The `_` isn't about visibility as it is in Python (and pre-Typescript JS), but that the field is never read. Without the `_`, clippy complains that the field is only written to but never read. Removing the field would be incorrect because the whole trick relies on the fact that `Parsed` outlives the node. Adding the `_` makes clippy shut up, because it indicates that this is an intentional "unused"